pkg/kubelet/cm/cpumanager/topology_hints.go

```diff
@@ -0,0 +1,179 @@
+		}
+	}
+	klog.Infof("[cpumanager] Topology Hints for pod: %v", cpuHints)
+	return cpuHints
+}
+
+func (m *manager) getReservedCPUs(topo *topology.CPUTopology) cpuset.CPUSet {
+	allCPUs := topo.CPUDetails.CPUs()
+	reservedCPUs := m.nodeAllocatableReservation[v1.ResourceCPU]
+	reservedCPUsFloat := float64(reservedCPUs.MilliValue()) / 1000
+	numReservedCPUs := int(math.Ceil(reservedCPUsFloat))
+	reserved, _ := takeByTopology(topo, allCPUs, numReservedCPUs)
+	klog.Infof("[cpumanager] Reserved CPUs: %v", reserved)
+	return reserved
+}
+
+func (m *manager) getAvailableCPUs(containerID string, reserved cpuset.CPUSet) cpuset.CPUSet {
+	var availableCPUs cpuset.CPUSet
+	if cset, ok := m.state.GetCPUSet(containerID); ok {
+		klog.Infof("[cpumanager] Reusing pre-assigned CPUSet: %v", cset)
+		availableCPUs = cset
+	} else {
+		// Otherwise, calculate the available CPUs from the topology.
+		availableCPUs = m.state.GetDefaultCPUSet().Difference(reserved)
+	}
+	return availableCPUs
+}
+
+func getCPUHintsAffinity(cpusPerSocket []int, requested int) []topologymanager.TopologyHint {
+	var cpuHints []topologymanager.TopologyHint
+	var totalCPUs int = 0
+	for i := 0; i < len(cpusPerSocket); i++ {
+		totalCPUs += cpusPerSocket[i]
+		if cpusPerSocket[i] >= requested {
+			mask, _ := socketmask.NewSocketMask(i)
+			cpuHints = append(cpuHints, topologymanager.TopologyHint{SocketAffinity: mask})
+		}
+	}
+	if totalCPUs >= requested {
+		crossSocketMask, crossSocket := buildCrossSocketMask(len(cpusPerSocket), cpusPerSocket)
+		if crossSocket {
+			cpuHints = append(cpuHints, topologymanager.TopologyHint{SocketAffinity: crossSocketMask})
+		}
+
+	}
+	return cpuHints
+}
+
+func getCPUHintsPreferred(cpuHints []topologymanager.TopologyHint, requested int, mostCPUs int) []topologymanager.TopologyHint {
+	// Check all hints for the hint with the lowest number of sockets (bestHint).
+	bestHint := cpuHints[0].SocketAffinity.Count()
+	for r := range cpuHints {
+		if cpuHints[r].SocketAffinity.Count() < bestHint {
+			bestHint = cpuHints[r].SocketAffinity.Count()
+		}
+	}
+	// Iterate over all hints and set 'Preferred: true/false' for each hint
+	for r := range cpuHints {
+		if cpuHints[r].SocketAffinity.Count() <= bestHint && requested <= mostCPUs {
+			// In the event that the hint has a lower or equal number of sockets to bestHint
+			// and one or more of those sockets could individually satisfy the number of CPUs
+			// requested, assuming all CPUs on the socket (minus reserved CPUs) are available,
+			// set hint to 'Preferred: true'.
+			cpuHints[r].Preferred = true
+		} else {
+			// In all other scenarios - 'Preferred: false'.
+			// Eg: If the number of sockets in the hint is larger than that of bestHint,
+			// then there is a more preferrable hint available.
+			// Eg: No socket could ever satisfy the request individually, assuming
+			// all CPUs on that socket (minus reserved CPUs) were available.
+			cpuHints[r].Preferred = false
```

Comment:
So I am not sure why cpu manager show give an hint which is preferred = false.  IMHO it should have preferred = true to all of it hints otherwise the following example won't work:
I have 4 VFs 2 VF from numa 0 and 2 VF from numa 1 device manager will return (1,1 preferred = true)
and I requested 2 cpus (I have 2 cpus free on each numa node) cpu manage hints should be [(1,0 preferred = true),(0,1 preferred = true), (1,1 preferred = true). because all 3 option are good. It will be up to the topology manager to find the preferred one of all.

Reply:
If hint providers return as follows:
cpu manager: [(1,0 preferred = true),(0,1 preferred = true), (1,1 preferred = false)]
device manager: [(1,1 preferred = true)]
topology manager will calculate [1,1 preferred = true]

As long as cpu manager returns a better hint eg [1,0] or [0,1] than cross-socket affinity [1,1], then the preferred status of the cross-socket affinity will not influence topology manager's decision. So it is set to false by default with no consequence (as above).

If the cross-socket affinity is indeed the best hint returned by cpu manager, then it will influence the decision of topology manager and should therefore be set to true.
[1,1 preferred = true].
However, in the outlying case that neither numa node could ever satisfy a request individually, but could satisfy a request collectively, the hints returned will be [1,1 preferred = false]. In this case the pod will fail in "strict" mode and pass in "preferred" mode.

Reply:
@nolancon, thank for the reply. 
The is the problem I see in the current logic. 
let say I have 4 vf, 2 VF on NUMA0 and 2 VF on NUMA1 and I have 2 cpus free on NUMA0 and 2 cpus free NUMA1 and we are using strict mode. 
Scenario A:
if a user request in the pod spec 4 VF and 2 cpus
the device plugin manager will return [(1,1 preferred = true)]
the cpu manger will return [(0,1 preferred=true), (1,0 preferred=true) (1,1 preferred = false)]
and we will get Topology Affinity Error.
Scenario B:
if a user request in the pod spec 4 VF and 4 cpus
the device plugin manager will return [(1,1 preferred = true)]
the cpu manager will return [ (1,1 preferred = true)]
A pod will be successfully created.
This seem to me as an  inconsistent behavior.  I would expect both Scenarios to create pod successfully.

Reply:
sorry I was mistaken in  Scenario A we will have a final topology hint (0,1  preferred=true) because on the bitwise and operation. I guess the question is what strict policy means? and if that what the user expected.   Do not  you think that the best hint in that case is (1,1)? so if you have VFs on 2 NUMA don't you want to have cpus on both NUMA has well?

Reply:
The 1,1 hint indicates that the resources can be taken from either node and delegates the allocation to the hint provider, So in the event there is more device requests or additional hint providers sending on hints we could see resources being allocated from either node - lets say we had another device plugin that could allocate a device from either node and decided to give from numa node 1 and lets say there was a huge page hint provider then it decided to give the resources from numa node 0. 

Whereas in the event that the hint was 0,1 then all hint providers would allocate to the same node where possible and those who can't spread across sockets - these leads to as many resources as possible coming from the one numa node.

Though I understand your case above, maybe we can think more on how to satisfy this case. 

And finally, the strict policy fails in the event that a preferred hint cannot be chosen for any particular resource - lets say in scenario A CPU Manager only returned (1,1 preferred = false) this would fail in strict policy.

Reply:
So maybe the name of the policy is confusing.
Trying to allocate many resources as possible for one numa seem like best effort policy.
Where strict can be one of the following: 
1. allocate all resources one numa 
2. allocate all resources if they have the same socket mask hint
what do you think?

Reply:
I agree that there is some confusion in the naming.

At present, there is really only one true "policy" that the `TopologyManager` implements, and that, I agree, this is a best-effort policy. 

The `preferred` vs. `strict`  distinction is actually more of a "rule" for deciding if we admit a pod or not when this best-effort policy is not able to find alignment that includes at least one NUMA node.

Maybe a better name for the policies would be "best-effort-preferred" and "best-effort-strict".

That being said, I am having a hard time picturing how to cleanly support the type of policy that you describe. At present, the SocketAffinity mask in a TopologyHint is designed to be interpreted as an actual _mask_ (i.e. resources can be pulled with affinity to _any_ of the sockets in the mask, not that they *must* be pulled from all of them).

If we attempted to implement a policy similar to what you describe, then each HintProvider would have to somehow know to treat the SocketAffinity mask in the TopologyHint differently (i.e. as a *must* pull resources with affinity to every socket in the mask instead of a pull from *any*).

Intertwining the policy set in the TopologyManager with how to interpret the bits in the SocketAffinity mask by a hint provider seems like a bad idea.

Moreover, I'm not sure how to handle the case where you have (for example) 2 devices that must be allocated cross-NUMA, but you have only requested a single CPU. In your scheme this would block pod admission, I guess, because a single CPU can never be allocated on both NUMA devices at once?

Maybe if we enumerate the exact semantics you would like to see supported, and how you envision the CPUManager and other hint providers doing their allocations based on these semantics, we can come up with a solution that is compatible with the current design.

Reply:
Ok so just to setup the terminology I will call best-effort the algorithm and preferred/strict the rule.

IMO, if the current design is to support just best effort algorithm in this phase then we don't need the preferred/ strict "rule". best effort should always succeed.   

I think that it will be useful that the topology manager  algorithm can be configurable and this can be a later enhancement to the current PR. 

Thought?

Reply:
I think I tend to agree.

I've summarized my thoughts around this as one of the points in my latest feedback document: https://docs.google.com/document/d/1g5Aqa0BncQGRedSJH0TJQWq3mw3VxpJ_ufO1qokJ1LE

I also implemented a small python program to compare and contrast the different "merging strategies" we've been discussing for the strict policy. Check it out here:
https://gist.github.com/klueska/02443e7ee37c6d225771ba6fc9c385fc